src/rt/array/append.d

```diff
@@ -0,0 +1,290 @@
+    auto f = cast(func)&impl;
+    return f(x, y);
+}
+
+
+/**
+ * TraceGC wrapper around $(REF _d_arrayappendT, rt,array,append).
+ * Bugs:
+ *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
+ */
+ref T[] _d_arrayappendTTrace(T)(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted
+{
+    ref T[] impl(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted
+    {
+        import rt.profilegc : accumulate;
+        import core.memory : GC;
+
+        static if (is(T == shared))
+            string name = shared(T[]).stringof;
+        else
+            string name = T[].stringof;
+
+        // FIXME: use rt.tracegc.accumulator when it is accessable in the future.
+        version (tracegc)
+        {
+            import core.stdc.stdio;
+
+            printf("%s file = '%.*s' line = %d function = '%.*s' type = %.*s\n",
+                __FUNCTION__.ptr,
+                file.length, file.ptr,
+                line,
+                funcname.length, funcname.ptr,
+                name.length, name.ptr
+            );
+        }
+
+        ulong currentlyAllocated = GC.stats().allocatedInCurrentThread;
+
+        scope(exit)
+        {
+            ulong size = GC.stats().allocatedInCurrentThread - currentlyAllocated;
+            if (size > 0)
+                accumulate(file, line, funcname, name, size);
+        }
+        return _d_arrayappendT(x, y);
+    }
+
+    alias func = ref T[] delegate(string file, int line, string funcname, return scope ref T[] x, scope T[] y) @trusted pure nothrow;
+    auto f = cast(func)&impl;
+    return f(file, line, funcname, x, y);
+}
+
+// This function cannot be imported because then rt.lifetime, and its dependencies needs to be included in the druntime
+// release. It also cannot be core.internal.externDFunc'd because that causes recursive template instantiation.
+
+/// See $(REF _d_arrayappendcTX, rt,lifetime,_d_arrayappendcTX)
+private extern (C) byte[] _d_arrayappendcTX(const TypeInfo ti, ref byte[] px, size_t n);
+
+/**
+ * Extend an array `px` by `n` elements.
+ * Caller must initialize those elements.
+ * Params:
+ *  px = the array that will be extended, taken as a reference
+ *  n = how many new elements to extend it with
+ * Returns:
+ *  The new value of `px`
+ * Bugs:
+ *   The safety level of this function is faked. It shows itself as `@trusted pure nothrow` to not break existing code.
+ */
+ref T[] _d_arrayappendcTX(T)(return scope ref T[] px, size_t n) @trusted
+{
+    ref T[] impl(return scope ref T[] px, size_t n) @trusted
+    {
+        // This template will be instanciated even in betterC due to the lowering code.
+        // So we need to disallow the calling of this function if that happens.
+        version (D_BetterC)
+            assert(0, "_d_arrayappendcTX does not support betterC!");
```

Comment:
What happens if you don't specifically have this check.  Won't the -betterC build just emit a generic error about not supporting `TypeInfo`?

Reply:
Also, if you do choose to keep this message, I don't think it should specifically mention `_d_arrayappendcTX` because that never appears in the user's code.  I suggest something worded more as "betterC does not support appending arrays".  

And I think that it can be a static assert at the beginning of the `_d_arrayappendcTX` template so the user gets a compile-time assertion failure instead of a runtime assertion failure.

Reply:
So without this check `test19561.d` will fail with `Error: TypeInfo cannot be used with -betterC`, even if that expression will be intercepted by the CTFE later.

and it cannot be a static assert because that breaks the lowering code.
```
import/rt/array/append.d(121): Error: static assert:  "_d_arrayappendcTX does not support betterC!"
import/rt/array/append.d(35):        instantiated from here: _d_arrayappendcTX!(immutable(char))
import/core/internal/arrayop.d(320):        instantiated from here: _d_arrayappendT!(immutable(char))
import/core/internal/arrayop.d(45):        instantiated from here: initScalarVecs!(int, "+=")
import/object.d(3696):        instantiated from here: arrayOp!(int[], int, "+=")
```